include/deal.II/fe/fe_poly_tensor.h

```diff
@@ -245,11 +248,33 @@ class FE_PolyTensor : public FiniteElement<dim, spacedim>
     // initialize fields only if really
     // necessary. otherwise, don't
     // allocate memory
+
+    bool update_transformed_shape_values = false;
+    bool update_transformed_shape_grads = false;
+    bool update_transformed_shape_hessian_tensors = false;
+
+    for (unsigned int i = 0; i < this->mapping_type.size(); ++i)
+      {
+        MappingType mapping_type = this->mapping_type[i];
+
+        if (mapping_type != mapping_none)
+          update_transformed_shape_values = true;
```

Comment:
In essence, what you're doing here is determine whether *any* of the elements of the vector are different from `mapping_none`. (And similarly in the next few lines.) You can write this more concisely in the following way (see https://en.cppreference.com/w/cpp/algorithm/find):
```
  const book update_transformed_shape_values = (std::find_if_not (this->mapping_type.begin(),
                                                                                                 this->mapping_type.begin(),
                                                                                                 [](const MappingType t) { return t != mapping_none; })
                                                                                   != this->mapping_type.end());
```

Reply:
This command is a mouthful. Let me see... so this command locates any entry between begin or end satisfies the inline function (what's the C++ word for that?) that returns true if the mapping type is not mapping_none. I assume the default return is `.end()` if nothing is found?

```
 const bool update_transformed_shape_values = (std::find_if (this->mapping_type.begin(),
                                                             this->mapping_type.end(),
                                                             [](const MappingType t) { return t != mapping_none; })
                                                                                   != this->mapping_type.end());
```

Should that be `.end()` on the second line? Also, if we use `return t != mapping_none`, shouldn't we use find_if instead of find_if_not?

And if I were to do this for transformed_shape_grads, would it look like this?

```
 const bool update_transformed_shape_values = (std::find_if (this->mapping_type.begin(),
                                                             this->mapping_type.end(),
                                                             [](const MappingType t) { return (t == mapping_raviart_thomas ||
                                                                                               t == mapping_piola ||
                                                                                               t == mapping_nedelec ||
                                                                                               t == mapping_contravariant); })
                                                                                   != this->mapping_type.end());
```

If so, then I think I understand how this works and I can implement it.

Reply:
Oh is that called a "lambda expression"? This is something I haven't explored much of before.

Reply:
Yes:
* That's a lambda function -- a function without a name defined in the place where you need it.
* And yes, correct, it should have been `.end()` in the second line.
* And yes, if `find_if` can't find an entry that satisfies the lambda function, then it returns `.end()`.

Reply:
https://en.wikipedia.org/wiki/Anonymous_function#C%2B%2B_%28since_C%2B%2B11%29